File: framec_tests/golang/transition_params/transition_params_test.go
package transition_params

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func (m *transitParamsStruct) log(msg string) {

	m.tape = append(m.tape, msg)
}

func TestEnter(t *testing.T) {
	sm := NewTransitParams()
	x := sm.(*transitParamsStruct)
	sm.Next()
	assert.Equal(t, []string{"hi A"}, x.tape)
}

func TestEnterAndExit(t *testing.T) {
	sm := NewTransitParams()
	x := sm.(*transitParamsStruct)
	sm.Next()
	x.tape = x.tape[:0]
	sm.Next()
	assert.Equal(t, []string{"bye A", "hi B", "42"}, x.tape)
	x.tape = x.tape[:0]
	sm.Next()
	assert.Equal(t, []string{"true", "bye B", "hi again A"}, x.tape)

}

func TestChangeState(t *testing.T) {
	sm := NewTransitParams()
	x := sm.(*transitParamsStruct)
	assert.Equal(t, TransitParamsState_Init, x._compartment_.State)
	sm.Change()
	assert.Equal(t, TransitParamsState_A, x._compartment_.State)
	sm.Change()
	assert.Equal(t, TransitParamsState_B, x._compartment_.State)
	sm.Change()
	assert.Equal(t, TransitParamsState_A, x._compartment_.State)
	assert.Empty(t, x.tape)

}

func TestChangeAndTransition(t *testing.T) {
	sm := NewTransitParams()
	x := sm.(*transitParamsStruct)
	sm.Change()
	assert.Equal(t, TransitParamsState_A, x._compartment_.State)
	assert.Empty(t, x.tape)
	sm.Next()
	assert.Equal(t, TransitParamsState_B, x._compartment_.State)
	assert.Equal(t, []string{"bye A", "hi B", "42"}, x.tape)
	x.tape = x.tape[:0]
	sm.Change()
	assert.Equal(t, TransitParamsState_A, x._compartment_.State)
	assert.Empty(t, x.tape)
	sm.Change()
	sm.Next()
	assert.Equal(t, TransitParamsState_A, x._compartment_.State)
	assert.Equal(t, []string{"true", "bye B", "hi again A"}, x.tape)

}

File: framec_tests/cpp/transition_params/transition_params_test.cpp
#include <gtest/gtest.h>
#include <vector>
#include <algorithm>
#include <string>
#include "transition_params.cpp"

class TransitParamsController : public TransitParams {
public:
    TransitParamsController() : TransitParams() {}
};

class TransitParamsControllerTest : public ::testing::Test {
protected:
    void SetUp() override {
        sm = new TransitParamsController();
    }

    void TearDown() override {
        delete sm;
    }

    TransitParamsController* sm;
};

TEST_F(TransitParamsControllerTest, TestEnter) {
    sm->Next();
    EXPECT_EQ(std::vector<std::string>{"hi A"}, sm->tape);
}

TEST_F(TransitParamsControllerTest, TestEnterAndExit) {
    sm->Next();
    sm->tape.clear();
    sm->Next();
    EXPECT_EQ((std::vector<std::string>{"bye A", "hi B", "42"}), sm->tape);
    sm->tape.clear();
    sm->Next();
    EXPECT_EQ((std::vector<std::string>{"true", "bye B", "hi again A"}), sm->tape);
}

TEST_F(TransitParamsControllerTest, TestChangeState) {
    EXPECT_EQ("0", sm->state_info());
    sm->Change();
    EXPECT_EQ("1", sm->state_info());
    sm->Change();
    EXPECT_EQ("2", sm->state_info());
    sm->Change();
    EXPECT_EQ("1", sm->state_info());
    EXPECT_TRUE(sm->tape.empty());
}

TEST_F(TransitParamsControllerTest, TestChangeAndTransition) {
    sm->Change();
    EXPECT_EQ("1", sm->state_info());
    EXPECT_TRUE(sm->tape.empty());
    sm->Next();
    EXPECT_EQ("2", sm->state_info());
    EXPECT_EQ((std::vector<std::string>{"bye A", "hi B", "42"}), sm->tape);
    sm->tape.clear();
    sm->Change();
    EXPECT_EQ("1", sm->state_info());
    EXPECT_TRUE(sm->tape.empty());
    sm->Change();
    sm->Next();
    EXPECT_EQ("1", sm->state_info());
    EXPECT_EQ((std::vector<std::string>{"true", "bye B", "hi again A"}), sm->tape);
}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
